### PR TITLE
[NEW] Onboarding Screen

### LINKED ISSUE #1053: [NEW] Create onboarding screen
Follow material [design guidelines](https://material.io/guidelines/growth-communications/onboarding.html#onboarding-onboarding-models)

### PR DESCRIPTION
**Pull request by divyanshub024**
@RocketChat/android

Closes #1053 

on2

### CODE REVIEW

File: app/src/main/java/chat/rocket/android/authentication/ui/AuthenticationActivity.kt [hunk 1]
```diff
@@ -23,18 +24,28 @@ class AuthenticationActivity : AppCompatActivity(), HasSupportFragmentInjector {
...
     val job = Job()
+    val PREF_FIRST_TIME_USER = "first_time_user"
 
     override fun onCreate(savedInstanceState: Bundle?) {
         AndroidInjection.inject(this)
         setContentView(R.layout.activity_authentication)
         setTheme(R.style.AuthenticationTheme)
         super.onCreate(savedInstanceState)
 
+        var isFirstTimeUser = Utils().readSharedSetting(this,PREF_FIRST_TIME_USER, "true")!!.toBoolean()
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/authentication/ui/AuthenticationActivity.kt]: Take a look at https://github.com/RocketChat/Rocket.Chat.Android/blob/develop/app/src/main/java/chat/rocket/android/infrastructure/LocalRepository.kt for saving this information and how we use it to check for a first time on https://github.com/RocketChat/Rocket.Chat.Android/blob/develop/app/src/main/java/chat/rocket/android/app/RocketChatApplication.kt#L102

File: app/src/main/java/chat/rocket/android/authentication/ui/AuthenticationActivity.kt [hunk 2]
```diff
@@ -23,18 +24,28 @@ class AuthenticationActivity : AppCompatActivity(), HasSupportFragmentInjector {
...
+
         launch(UI + job) {
             val newServer = intent.getBooleanExtra(INTENT_ADD_NEW_SERVER, false)
             presenter.loadCredentials(newServer) { authenticated ->
                 if (!authenticated) {
                     showServerInput(savedInstanceState)
+                    if (isFirstTimeUser)
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/authentication/ui/AuthenticationActivity.kt]: This is getting more complicated, we should move all the logic into the Presenter, and make it call the right `view`method

File: app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_onboarding)
+
+        val adapter = OnboardingPagerAdapter(supportFragmentManager)
+        view_pager.adapter = adapter
+        pageIndicator.initViewPager(view_pager)
+
+        button_next.setOnClickListener{
+            if (view_pager.currentItem != adapter.count-1){
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt]: Missing space in `adapter.count-1` shoud be `adapter.count - 1`. The same happens on a lot of places on this PR.

File: app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        view_pager.addOnPageChangeListener(object: ViewPager.OnPageChangeListener{
+
+            override fun onPageScrollStateChanged(state: Int) {}
+
+            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
+
+                val topRightX = image_mobile.x + image_mobile.width
+                val topRightY = image_mobile.y
+                when(position){
+                    0->{
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt]: should be `0 -> {`

File: app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                    }
+                    2 ->{
+                        button_skip.setVisible(false)
+                        button_next.text = resources.getString(R.string.action_done)
+                    }
+                }
+            }
+
+            override fun onPageSelected(position: Int) {}
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt]: unneeded blank line

File: app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        })
+
+        view_pager.setPageTransformer(false,PageTransformerPresenter())
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt]: unneeded blank line

File: app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    val Int.px: Int
+        get() = (this * Resources.getSystem().displayMetrics.density).toInt()
+
+    override fun onBackPressed() {
+        if (view_pager.currentItem !=0){
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt]: ` != 0`

File: app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            view_pager.currentItem = view_pager.currentItem - 1
+        }else{
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt]: `} else {`

File: app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingFragment1.kt [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import chat.rocket.android.onboarding.presentation.PageTransformerView
+import kotlinx.android.synthetic.main.fragment_onboarding_fragment.*
+import kotlinx.android.synthetic.main.fragment_onboarding_fragment.view.*
+
+class OnboardingFragment1 : Fragment(), PageTransformerView {
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
+                              savedInstanceState: Bundle?): View? {
+
+        val view =  inflater.inflate(R.layout.fragment_onboarding_fragment, container, false)
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingFragment1.kt]: There is an extra space after the `= `

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    private var startedSettleNormalOffset: Float = 0.toFloat()
+    private var startedSettlePageOffset: Float = 0.toFloat()
+
+    private val desiredHeight: Int
+        get() = paddingTop + paddingBottom + circleRadius * 2
+
+    private val desiredWidth: Int
+        get() =
+            paddingLeft + paddingRight + circleRadius * 2 * circleCount + (circleCount - 1) * circlePadding
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+                Gravity.RIGHT, GravityCompat.END -> return measuredWidth - paddingRight - allCircleWidth
+                Gravity.CENTER -> return measuredWidth / 2 - allCircleWidth / 2
+                else -> return measuredWidth / 2 - allCircleWidth / 2
+            }
+        }
+
+    val allCircleWidth: Int
+        get() = circleRadius * 2 * circleCount + (circleCount - 1) * circlePadding
+
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        init(attrs)
+    }
+
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
+        init(attrs)
+    }
+
+    private fun init(attrs: AttributeSet?) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+            activeColor = typedArray.getColor(R.styleable.PageIndicator_indicatorActiveColor, activeColor)
+            mGravity = typedArray.getInt(R.styleable.PageIndicator_android_gravity, gravity)
+        }
+
+        activePaint = Paint()
+        activePaint!!.color = activeColor
+
+        inActivePaint = Paint()
+        inActivePaint!!.color = inactiveColor
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+        mViewPager = viewPager
+        mViewPager!!.addOnPageChangeListener(this)
+
+        circleCount = viewPager.adapter!!.count
+        mCurrentDragPage = viewPager.currentItem
+        invalidate()
+    }
+
+    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+        val height = View.resolveSize(desiredHeight, heightMeasureSpec)
+        val width = View.resolveSize(desiredWidth, widthMeasureSpec)
+
+        setMeasuredDimension(width, height)
+    }
+
+    override fun onDraw(canvas: Canvas) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+        // draw behind circles
+        for (i in 0 until circleCount) {
+            val center = getCircleCenter(i)
+            canvas.drawCircle(center, (paddingTop + circleRadius).toFloat(), circleRadius.toFloat(), inActivePaint!!)
+        }
+
+        drawRect(canvas)
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    private fun drawRect(canvas: Canvas) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+        if (mViewPager == null)
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: You can use a single if statement here, `if (mViewPager == null || mViewPager!!.adapter == null || mViewPager!!.adapter!!.count == 0)`

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+
+        val top = paddingTop.toFloat()
+        val bottom = top + circleRadius * 2
+
+        val moveDistance = (circleRadius * 2 + circlePadding).toFloat()
+        val isDragForward = mSelectedPage - mCurrentDragPage < 1
+
+        val relativePageOffset = if (isDragForward) mPageOffSet else 1.0f - mPageOffSet
+        currentRelativePageOffset = relativePageOffset
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt:141]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+        val left = if (isDragForward) circleCenter - circleRadius + normal else circleCenter - circleRadius.toFloat() - large
+        val right = if (isDragForward) circleCenter + circleRadius.toFloat() + large else circleCenter + circleRadius - normal
+
+        val rectF = RectF(left, top, right, bottom)
+        canvas.drawRoundRect(rectF, circleRadius.toFloat(), circleRadius.toFloat(), activePaint!!)
+    }
+
+    private fun getCircleCenter(position: Int): Float {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    private fun getCirclePadding(position: Int): Float {
+        return (circlePadding * position + circleRadius * 2 * position).toFloat()
+    }
+
+    private fun mapValue(value: Float, a1: Float, a2: Float, b1: Float, b2: Float): Float {
+        return b1 + (value - a1) * (b2 - b1) / (a2 - a1)
+    }
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt:174]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+
+    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+        mCurrentDragPage = position
+        mPageOffSet = positionOffset
+        postInvalidate()
+    }
+
+    override fun onPageSelected(position: Int) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    override fun onPageScrollStateChanged(state: Int) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+    override fun onDetachedFromWindow() {
+        if (mViewPager != null)
+            mViewPager!!.removeOnPageChangeListener(this)
+        //finish modified
+        super.onDetachedFromWindow()
+    }
+
+    companion object {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: blank line

File: app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+        //setted as final
+        private val OFFSET_MULTIPLIER_DRAG = 1.2f
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/widget/pageIndicator/PageIndicator.kt]: use `private const val`

File: app/src/main/java/chat/rocket/android/onboarding/presentation/PageTransformerPresenter.kt [hunk 27]
```diff
@@ -0,0 +1,24 @@
+package chat.rocket.android.onboarding.presentation
+
+import android.support.v4.view.ViewPager
+import android.view.View
+
+class PageTransformerPresenter : ViewPager.PageTransformer {
+
+    override fun transformPage(page: View, position: Float) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/presentation/PageTransformerPresenter.kt]: blank line

File: app/src/main/java/chat/rocket/android/onboarding/presentation/PageTransformerPresenter.kt [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+        if (page.tag is PageTransformerView) {
+
```
Comment (luciofm) [inline app/src/main/java/chat/rocket/android/onboarding/presentation/PageTransformerPresenter.kt]: blank line

**Review by luciofm — CHANGES_REQUESTED** [review]
Added some comments

File: app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt (hunk 3 above)
Comment (Shailesh351) [reply app/src/main/java/chat/rocket/android/onboarding/ui/OnboardingActivity.kt]: @divyanshub024 You can try `ctrl + alt + shift + L` for formating selected text or whole file in android studio :smile:

**Comment by leonardoaramaki** [comment]
Could you please solve the conflicts?

File: app/src/main/java/chat/rocket/android/authentication/ui/AuthenticationActivity.kt (hunk 1 above)
Comment (divyanshub024) [reply app/src/main/java/chat/rocket/android/authentication/ui/AuthenticationActivity.kt]: @luciofm can you tell me how to check for first time user. The code has been updated.

**Comment by divyanshub024** [comment]
@luciofm can you please tell me how to check for the first time user.

**Comment by divyanshub024** [comment]
Closing since that layout will be change very soon

**Pull request closed without merging**